### ISSUE TITLE
Menu bar not accessible in Ubuntu/Unity

### ISSUE THREAD

**Issue by mihaiile** [keepassxreboot/keepassxc#271]
The menu bar is no longer accessible if the **Hide Window to system tray when minimized** is selected.

## Expected Behavior

The menu bar should be accessible after minimising to tray and showing back the window

## Current Behavior

Menu bar is not accessible/visible after bringing back the window from system tray

## Steps to Reproduce (for bugs)
1. Open application, from menu go to Tools->Settings
2. Under General check **Show a system tray icon** and **Hide Window to system tray when minimized**
3. Save settings and minimise the application.
4. From the tray choose **Toggle window** option to show the application again.
5. Try to access menu bar again.

## Context

I would say this bug is quite critical since if one has checked also the setting **Minimize window at application startup** as I have, there is no way you can ever again access the menu bar.
The only option I was able to find is go to config file and change the **MinimizeOnStartup=true** entry.

## Your Environment

**Extra Issue: It is impossible to copy text from Help->About**
* KeePassXC version/commit used: 2.1.1 (KeePassXC-2.1.1-x86_64.AppImage)
* Qt version: 5.8.0
* Operating System and version: Ubuntu 16.10 64bit

**Comment by phoerious** [comment]
I spent a couple of hours now hunting down the bug and evaluating ways to fix it.
Qt supports native global application menus via D-Bus on Linux starting with 5.7.0. The new AppImage release is built on 5.8.0 instead of 5.2.1 as before. Unfortunately, the Ubuntu un-registers hidden windows and Qt doesn't re-register the menu once the window is shown again.

To make matters worse, the routines for registering the global menu are not part of the public Qt API and I cannot trigger them manually, therefore. I reported this issue to Qt https://bugreports.qt.io/browse/QTBUG-58723, but I'm not sure if and when it will be fixed.

For the time being, I see no proper solution except disabling the global menu or adding a lot of custom code to re-implement the D-Bus interface.

**Comment by phoerious** [comment]
I committed a patch with a workaround to the hotfix/271-dbus-menu-ubuntu. Would you be able to test it, please?

**Comment by mihaiile** [comment]
Unfortunately it still does not work on Ubuntu 16.10
Is there a verbose mode I could launch it to give more feedback?

**Comment by phoerious** [comment]
I tested it on 16.04 and it works for me now. Make sure you have checked out the correct branch: https://github.com/keepassxreboot/keepassxc/pull/276

When you run
```
dbus-monitor interface=com.canonical.AppMenu.Registrar
```
in a terminal, you should see a message like this every time you restore from the tray:
```
method call time=1486649954.141331 sender=:1.180 -> destination=com.canonical.AppMenu.Registrar serial=326 path=/com/canonical/AppMenu/Registrar; interface=com.canonical.AppMenu.Registrar; member=RegisterWindow
   uint32 136314885
   object path "/MenuBar/1"
```

**Comment by phoerious** [comment]
Since you are using Ubuntu 16.10, I guess your Qt version is also too old. If you don't seen any such D-Bus messages, could you go to src/MainWindow.cpp, line 744 and replace the line
```
#if defined(Q_OS_LINUX) && ! defined(QT_NO_DBUS) && QT_VERSION >= QT_VERSION_CHECK(5, 7, 0)
```
with
```
#if defined(Q_OS_LINUX) && ! defined(QT_NO_DBUS)
```
and then compile and test again?

**Comment by mihaiile** [comment]
This is my environment for the application:
![screenshot from 2017-02-09 16-18-12](https://cloud.githubusercontent.com/assets/3273985/22789527/70114bd2-eee3-11e6-942b-83ed8dd667bb.png)

This is what I get from dbus-monitor when I:
-launch application
-minimise to tray
-toggle window from tray
-close application

```
method call time=1486653557.696553 sender=:1.474 -> destination=com.canonical.AppMenu.Registrar serial=10 path=/com/canonical/AppMenu/Registrar; interface=com.canonical.AppMenu.Registrar; member=RegisterWindow
   uint32 62914564
   object path "/MenuBar/1"
signal time=1486653557.697003 sender=:1.79 -> destination=(null destination) serial=123057 path=/com/canonical/AppMenu/Registrar; interface=com.canonical.AppMenu.Registrar; member=WindowRegistered
   uint32 62914564
   string ":1.474"
   object path "/MenuBar/1"
method call time=1486653557.795928 sender=:1.74 -> destination=com.canonical.AppMenu.Registrar serial=3529 path=/com/canonical/AppMenu/Registrar; interface=com.canonical.AppMenu.Registrar; member=GetMenuForWindow
   uint32 62914564
signal time=1486653563.665233 sender=:1.79 -> destination=(null destination) serial=123149 path=/com/canonical/AppMenu/Registrar; interface=com.canonical.AppMenu.Registrar; member=WindowUnregistered
   uint32 62914564
method call time=1486653566.581421 sender=:1.74 -> destination=com.canonical.AppMenu.Registrar serial=3564 path=/com/canonical/AppMenu/Registrar; interface=com.canonical.AppMenu.Registrar; member=GetMenuForWindow
   uint32 62914564
method call time=1486653569.090567 sender=:1.474 -> destination=com.canonical.AppMenu.Registrar serial=214 path=/com/canonical/AppMenu/Registrar; interface=com.canonical.AppMenu.Registrar; member=UnregisterWindow
   uint32 62914564
signal time=1486653569.091310 sender=:1.79 -> destination=(null destination) serial=123223 path=/com/canonical/AppMenu/Registrar; interface=com.canonical.AppMenu.Registrar; member=WindowUnregistered
   uint32 62914564
signal time=1486653569.128552 sender=:1.79 -> destination=(null destination) serial=123227 path=/com/canonical/AppMenu/Registrar; interface=com.canonical.AppMenu.Registrar; member=WindowUnregistered
   uint32 62914564

```

**Comment by phoerious** [comment]
As I suspected: you are using Qt 5.6 which is the default that comes with Ubuntu 16.10.
I compiled my test case into an AppImage with Qt 5.8 and tested that.

Two things I would like you to test:

1. see my comment above (remove the check for Qt >= 5.7)
2. even if that works, could you temporarily uninstall appmenu-qt5 and test whether KeePassXC still uses a global menu?

**Comment by mihaiile** [comment]
Yes, using
```
#if defined(Q_OS_LINUX) && ! defined(QT_NO_DBUS)
```
works as expected.

**Comment by phoerious** [comment]
Alright, then I need to remove that check. Could you still test 2, please?
I just want to be sure that it's appmenu-qt5 that is providing the global menu in your case.

**Comment by mihaiile** [comment]
It seems 16.10 does not come default with appmenu-qt5 so that should be fine.

This is what I get when searching for it:
![screenshot from 2017-02-09 16-29-43](https://cloud.githubusercontent.com/assets/3273985/22789967/0117c862-eee5-11e6-9299-61e4bae65b8b.png)

**Comment by phoerious** [comment]
Weird. Could you try uninstalling appmenu-qt then? This might be Qt4, but I'm not sure.

**Comment by mihaiile** [comment]
I did uninstall the appmenu-qt and rand the application just fine, then installed appmenu-qt5 and also ran fine, so to me with ot without appmenu-qt or appmenu-qt or both installed and uninstalled it worked.
Note: I did not restart the system or logout/login after install/reinstall of the appmenu-qt packages.

**Comment by phoerious** [comment]
Hmm. I wonder what provides the global menu for you then. But maybe Canonical backported native DBus menus to Qt 5.6.x.

Anyway, thanks a lot for testing and reporting this in the first place!

**Comment by phoerious** [comment]
Bugfix for this was just merged into Qt 5.9.0 RC.
